In [1]:
import sys
import os 

sys.path.append(os.path.abspath(".."))

# LOADING THE DATASET FOR ANALYSIS

In [2]:
from src.data.load_data import load_raw_data

df = load_raw_data("../data/raw/Online Retail.xlsx")

df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


# MISSING (NULL) VALUE ANALYSIS

Column-wise misisng analysis

In [3]:
df.isnull().mean()

InvoiceNo      0.000000
StockCode      0.000000
Description    0.002683
Quantity       0.000000
InvoiceDate    0.000000
UnitPrice      0.000000
CustomerID     0.249267
Country        0.000000
dtype: float64

* CustomerID: ~25.0% missing  
* Description: ~0.26% missing  

In [5]:
df[df["CustomerID"].isnull()].head()

df[df["CustomerID"].isnull()].describe()

df[df["CustomerID"].isnull()]["Country"].value_counts()

Country
United Kingdom    133600
EIRE                 711
Hong Kong            288
Unspecified          202
Switzerland          125
France                66
Israel                47
Portugal              39
Bahrain                2
Name: count, dtype: int64

Rows with missing CustomerID appears to represent transactions without a  
unique customer identifier. These may corresponds to guest users or  
anonymous purchases. Hence, these rows are unsuitable for customer-level modeling.  

In [58]:
customerID_missing = df["CustomerID"].isnull().mean()*100
description_missing = df["Description"].isnull().mean()*100

print(f"Total Data points affected due to missing values of CustomerIDs: {customerID_missing}")
print(f"Total Data points affected due to missing Descriptions: {description_missing}")

Total Data points affected due to missing values of CustomerIDs: 24.926694334288598
Total Data points affected due to missing Descriptions: 0.2683107311375157


**COMBINED IMPACT :**  

In [6]:
((df["CustomerID"].isnull()) | (df["Description"].isnull())).mean()*100

np.float64(24.926694334288598)

Approximately 24.93% of the dataset contains missing values in either  
CustomerID or Description, indicating that 24.93% of the data points from the datasets  
may not be usable for certian analysis.

**Final Insight**
* The missing CustomerID values are not random but structurally missing,  
  indicating a specific class of transactions rather than data curruption.  
* CustomerID missing values represent transactions without identfiable customers.  
* These records cannot be used for customer-level feature engineering.  
* Description missing values are minimal.  
* A significant portion (~25%) of the dataset is affected by missing  
  CustomerID, which influence downstream modeling decisions. 